### PR TITLE
Adds an option recursive in append_values and spread_values

### PR DESCRIPTION
**Pull request by vats-div**

### CODE REVIEW

File: R/append_values.r [hunk 1]
```diff
@@ -59,10 +70,10 @@ append_values_factory <- function(type, as.value) {
 
 #' Unlists while preserving NULLs and only unlisting lists with one value
 #' @param l a list that we want to unlist
-my_unlist <- function(l) {
+my_unlist <- function(l, recursive = FALSE) {
```
Comment (jeremystan) [inline R/append_values.r]: @vats-div need to document recursive in this function to pass devtools::check() even though it isn't external

File: tests/testthat/test-append_values.r [hunk 2]
```diff
@@ -182,6 +182,31 @@ test_that("correctly handles append when trying to append an array", {
...
 
+test_that("recursive works as expected", {
+
+   data <- '{"item1": {"price" : 30}, "item2" : 40, "item3" : 30}' %>% gather_keys 
+   expected_na <- c(NA_real_, 40, 30)
+   expected_val <- c(30, 40, 30)
+
+   expect_identical((data %>% append_values_number())$number, expected_na)
+   expect_identical((data %>% append_values_number(recursive=TRUE))$number, expected_val)
+   expect_identical((data %>% append_values_number(force=FALSE, recursive=TRUE))$number,
```
Comment (jeremystan) [inline tests/testthat/test-append_values.r]: I don't think this is intuitive -- how recursive = TRUE behaves is a function of what force is set to? If you don't want recursive to be used unless force is set to TRUE, then I would validate that in the input and throw an error if it's not used correctly.

File: tests/testthat/test-spread_values.r [hunk 3]
```diff
@@ -72,48 +84,58 @@ context("spread_values")
...
+          document.id = 1L,
+          name = "bob",
+          age = 32,
+          customer = TRUE,
+          stringsAsFactors = FALSE
+        ), list(fromJSON(json)))
     
     expect_identical(
-      json %>%
+      json %>% as.tbl_json %>%
```
Comment (jeremystan) [inline tests/testthat/test-spread_values.r]: You don't need an explicity `as.tbl_json` anymore, I removed it from all tests unless it was strictly necessary.

File: tests/testthat/test-spread_values.r [hunk 4]
```diff
@@ -26,6 +29,9 @@ test_that("handles missing input properly", {
     expect_identical(jstring("name")(json),
       c("bob", "", NA_character_, NA_character_, NA_character_)
     )
+    expect_identical(jstring("name", recursive=TRUE)(json),
```
Comment (jeremystan) [inline tests/testthat/test-spread_values.r]: I'm not sure we need to add a recursive=TRUE mirror for every basic test? It seems unwieldy to maintain over time. If you want, you could write a wrapper that tests with and without recursion, and then use it everywhere, but that seems over-engineered.

File: R/spread_values.r [hunk 5]
```diff
@@ -62,6 +67,8 @@ jfactory <- function(na.value, conversion.function) {
 #' 
 #' @name jfunctions
 #' @param ... the path to follow
+#' @param recursive logical indicating whether second level and beyond objects
+#'        should be extracted.  Only works when there exists one a single value
```
Comment (jeremystan) [inline R/spread_values.r]: This is incomplete?

File: R/spread_values.r [hunk 6]
```diff
@@ -41,18 +41,23 @@ spread_values <- function(x, ...) {
...
+
     # Return a closure to deal with JSON lists
     function(json) {
       data <- list_path(json, path)
       data <- replace_nulls(data, na.value)
-      conversion.function(data)
+      if (recursive == FALSE) {
+         conversion.function(data)
+      } else {
+         vapply(data, function(d) conversion.function(unlist(d)),
```
Comment (jeremystan) [inline R/spread_values.r]: Should this use `tidyjson:::my_unlist` instead?

Comment (vats-div) [inline R/spread_values.r]: not needed since replace_nulls removes nulls.

File: R/append_values.r [hunk 7]
```diff
@@ -27,9 +30,11 @@ NULL
 #' @param as.value function to force coercion to numeric, string, or logical
 append_values_factory <- function(type, as.value) {
   
-  function(x, column.name = type, force=TRUE) {
+  function(x, column.name = type, force=TRUE, recursive=FALSE) {
```
Comment (jeremystan) [inline R/append_values.r]: Minor formatting - add a space around equal signs

File: R/append_values.r [hunk 8]
```diff
@@ -46,7 +51,15 @@ append_values_factory <- function(type, as.value) {
     if (!force) { 
        x[column.name] <- append_values_type(json, type) %>% as.value
     } else {
-       new_val <- my_unlist(json) %>% as.value
+       new_val <- my_unlist(json, recursive)
+
+       # if new_val is a list and recursive = FALSE, then
+       # need to identify values with a name and change to NA
+       if (is.list(new_val) & recursive == FALSE) {
```
Comment (jeremystan) [inline R/append_values.r]: This should be && (for boolean) not & (for vector-wise)

File: R/append_values.r [hunk 9]
```diff
@@ -46,7 +51,15 @@ append_values_factory <- function(type, as.value) {
...
+          loc <- (names(new_val) != "")
```
Comment (jeremystan) [inline R/append_values.r]: don't need the surrounding parentheses. should you check for `NULL` in addition to `""`? Is `""` the same as `NA_character_`?

File: R/append_values.r [hunk 10]
```diff
@@ -46,7 +51,15 @@ append_values_factory <- function(type, as.value) {
...
+          new_val[loc] <- NA
```
Comment (jeremystan) [inline R/append_values.r]: Can we specify the NA type explicitly? I think I do this elsewhere within a function factory.

File: R/append_values.r [hunk 11]
```diff
@@ -46,7 +51,15 @@ append_values_factory <- function(type, as.value) {
...
+       }
+       new_val <- new_val %>% as.value
```
Comment (jeremystan) [inline R/append_values.r]: where is `as.value` from? It's not in base R or dplyr?

File: R/append_values.r (hunk 9 above)
Comment (vats-div) [inline R/append_values.r]: This check is because of the way jsonlite parses the json.

**Comment by jeremystan** [comment]
+1

**Pull request merged**